ExampleCodes/Particles/ElectrostaticPIC/src/particles/ElectrostaticParticleContainer.cpp

```diff
@@ -0,0 +1,242 @@
+#include "field_solver/FieldSolver.H"
+
+#include "particles/ElectrostaticParticleContainer.H"
+#include "particles/pusher/ParticlePusher_K.H"
+#include "particles/field_gather/FieldGather_K.H"
+#include "particles/deposition/ChargeDeposition_K.H"
+
+#include <iomanip>
+
+using namespace amrex;
+
+void ElectrostaticParticleContainer::InitParticles () {
+
+    if ( ParallelDescriptor::MyProc() == ParallelDescriptor::IOProcessorNumber() ) {
+
+        ParticleType p;
+
+        p.id()   = ParticleType::NextID();
+        p.cpu()  = ParallelDescriptor::MyProc();
+
+        p.pos(0) = -2.5e-6;
+        p.pos(1) =  0.0;
+
+        std::array<ParticleReal,PIdx::nattribs> attribs;
+        attribs[PIdx::w]  = 1.0;
+        attribs[PIdx::vx] = 0.0;
+        attribs[PIdx::vy] = 0.0;
+        attribs[PIdx::Ex] = 0.0;
+        attribs[PIdx::Ey] = 0.0;
+
+        // Add to level 0, grid 0, and tile 0
+        // Redistribute() will move it to the proper place.
+        std::pair<int,int> key {0,0};
+        auto& particle_tile = GetParticles(0)[key];
+
+        particle_tile.push_back(p);
+        particle_tile.push_back_real(attribs);
+
+    }
+
+    Redistribute();
+}
+
+void
+ElectrostaticParticleContainer::DepositCharge (const Vector<MultiFab*>& rho) {
+
+    int num_levels = rho.size();
+    int finest_level = num_levels - 1;
+
+    // each level deposits it's own particles
+    const int ng = rho[0]->nGrow();
+    for (int lev = 0; lev < num_levels; ++lev) {
+        rho[lev]->setVal(0.0, ng);
+        const auto& gm = m_gdb->Geom(lev);
+        auto plo = gm.ProbLoArray();
+        auto dxi = gm.InvCellSizeArray();
+        for (MyParIter pti(*this, lev); pti.isValid(); ++pti) {
+            const Long np  = pti.numParticles();
+            const auto& wp = pti.GetAttribs(PIdx::w);
+            const auto& particles = pti.GetArrayOfStructs();
+
+            amrex::Real q = this->charge;
+            auto rhoarr = (*rho[lev])[pti].array();
+            const auto wp_ptr = wp.data();
+            const auto p_ptr = particles().data();
+            amrex::ParallelFor(np, [=] AMREX_GPU_DEVICE (int i) noexcept {
+                          deposit_cic(p_ptr[i], wp_ptr[i], q, rhoarr, plo, dxi);
+                                   });
+        }
+
+        rho[lev]->SumBoundary(gm.periodicity());
+    }
+
+    // now we average down fine to crse
+    IntVect ratio(D_DECL(2, 2, 2));  // FIXME
```

Comment:
FIXME?

Reply:
FIXME because the ref ratio should not be hard-coded here. ParticleContainers actually know the ref_ratio, so I should use the member data here.